File: frontend/src/utils/heiken.ts
import type { Candle } from '../types';

/**
 * Converts standard OHLC candles into Heiken Ashi candles.
 *
 * HA Close = (Open + High + Low + Close) / 4
 * HA Open  = (Previous HA Open + Previous HA Close) / 2  (first bar: (O+C)/2)
 * HA High  = max(High, HA Open, HA Close)
 * HA Low   = min(Low, HA Open, HA Close)
 *
 * Runs once, in a single O(n) pass, when a dataset is loaded. Never
 * recomputed during replay — the replay engine only reveals/hides values
 * that already exist in this array.
 */
export function toHeikenAshi(candles: Candle[]): Candle[] {
  const result: Candle[] = new Array(candles.length);
  let prevHaOpen = 0;
  let prevHaClose = 0;

  for (let i = 0; i < candles.length; i++) {
    const c = candles[i];
    const haClose = (c.open + c.high + c.low + c.close) / 4;
    const haOpen = i === 0 ? (c.open + c.close) / 2 : (prevHaOpen + prevHaClose) / 2;
    const haHigh = Math.max(c.high, haOpen, haClose);
    const haLow = Math.min(c.low, haOpen, haClose);

    result[i] = {
      time: c.time,
      open: haOpen,
      high: haHigh,
      low: haLow,
      close: haClose,
      volume: c.volume,
      oi: c.oi,
      symbol: c.symbol,
      expiry: c.expiry,
      strike: c.strike,
      exchange: c.exchange,
    };

    prevHaOpen = haOpen;
    prevHaClose = haClose;
  }

  return result;
}

File: backend/heiken.py
"""
Heiken Ashi candle generation.

HA Close = (Open + High + Low + Close) / 4
HA Open  = (Previous HA Open + Previous HA Close) / 2   (first candle: (O+C)/2)
HA High  = max(High, HA Open, HA Close)
HA Low   = min(Low, HA Open, HA Close)

Computed once, in a single O(n) pass. The frontend also implements this
(see src/utils/heiken.ts) so replay stays fully client-side; this module
exists so the backend can optionally serve pre-computed HA data too.
"""
from __future__ import annotations

from models import Candle


def to_heiken_ashi(candles: list[Candle]) -> list[Candle]:
    ha_candles: list[Candle] = []
    prev_ha_open: float | None = None
    prev_ha_close: float | None = None

    for c in candles:
        ha_close = (c.open + c.high + c.low + c.close) / 4.0
        if prev_ha_open is None or prev_ha_close is None:
            ha_open = (c.open + c.close) / 2.0
        else:
            ha_open = (prev_ha_open + prev_ha_close) / 2.0
        ha_high = max(c.high, ha_open, ha_close)
        ha_low = min(c.low, ha_open, ha_close)

        ha_candles.append(
            Candle(
                time=c.time,
                open=ha_open,
                high=ha_high,
                low=ha_low,
                close=ha_close,
                volume=c.volume,
                oi=c.oi,
                symbol=c.symbol,
                expiry=c.expiry,
                strike=c.strike,
                exchange=c.exchange,
            )
        )
        prev_ha_open, prev_ha_close = ha_open, ha_close

    return ha_candles
